Add KEP for Elastic Jobsets (#1147)
kubernetes-sigs/jobset: keps/463-ElasticJobsets/README.md

Kueue cannot scale JobSets today. With elastic support, Kueue scales based on cluster utilization:

```
Cluster 100 GPUs @ 80% → Kueue scales: replicas 20→25
Cluster 100 GPUs @ 95% → Kueue scales: replicas 25→20
```

### Risks and Mitigations

- **Resource Leakage:** Deleting a replica might leave orphaned pods if the child Job is not cleaned up properly.
  - *Mitigation:* The controller will rely on OwnerReferences and the standard Job deletion lifecycle to ensure all pods are terminated.
- **Index Fragmentation:** Scaling down and then up could lead to gaps in naming or indexing.
  - *Mitigation:* The controller will always scale down by removing the highest-indexed Jobs and scale up by appending new indices sequentially.

## Design Details

### API Details

The validation webhook for `jobset.x-k8s.io/v1alpha2` will be modified to allow updates to `spec.replicatedJobs[i].replicas`.

```golang

type ReplicatedJob struct {
	// Name of the ReplicatedJob.
	Name string `json:"name"`

	// Replicas is the number of desired Jobs for this ReplicatedJob.
	// This field is changed from immutable to mutable.
	Replicas int32 `json:"replicas"`

> 	// ... (Template and other fields remain immutable)

Author: RIght. The JobSet webhook currently enforces immutability on the entire template. I've updated the KEP to explicitly relax the validation for `parallelism` and `completions` in the `ReplicatedJob` template, allowing us to propagate those patches to the underlying Jobs.

Please review and let me know if this is acceptable.

title can help communicate what the KEP is and should be considered as part of
any review.
-->

<!--
A table of contents is helpful for quickly jumping to sections of a KEP and for
highlighting any additional information provided beyond the standard KEP
template.

Ensure the TOC is wrapped with
  <code>&lt;!-- toc --&rt;&lt;!-- /toc --&rt;</code>
tags, and then generate with `hack/update-toc.sh`.
-->

<!-- toc -->
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [User Stories](#user-stories)
    - [Story 1](#story-1)
    - [Story 2](#story-2)
    - [Story 3](#story-3)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [API Details](#api-details)
  - [Implementation](#implementation)
  - [Failure Handling Priority](#failure-handling-priority)
  - [Defaulting/Validation](#defaultingvalidation)
  - [Test Plan](#test-plan)
    - [Unit Tests](#unit-tests)
    - [Integration Tests](#integration-tests)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

> This KEP outlines the proposal to support "elasticity" within JobSets by making the `replicas` field of a `ReplicatedJob` mutable. This allows external controllers or users to scale the number of Job replicas in a JobSet up or down dynamically during its execution.

Reviewer: What about Job's `parallelism` and `completions`? Shall we also allow to modify that field?
Those fields are immutable now in JobSet: https://github.com/kubernetes-sigs/jobset/blob/main/pkg/webhooks/jobset_webhook.go#L350C24-L350C46

Reviewer: Found it later in the doc, so we just need to update Summary.

## Motivation

Currently, most fields in a JobSet are immutable after creation. This poses a challenge for distributed training workloads that are designed to be elastic. For example, PyTorch Elastic (Torchrun) can handle a varying number of nodes. If a JobSet starts with 4 replicas and one node fails or is preempted, the JobSet should ideally scale down to 3 replicas and allow the training to continue, rather than failing the entire workload.

### Goals

> 

Author: Right. Added that

- Support scaling the number of replicas in a `ReplicatedJob` after the JobSet has been created.
- Ensure the JobSet controller correctly manages the lifecycle (creation/deletion) of child Jobs based on the updated replica count.

> - Maintain the stability of existing JobSet features like `StartupPolicy` and `ExclusivePlacement` during scaling events.

Reviewer: ```suggestion
- Maintain the stability of existing JobSet features like `DependsOn` and `ExclusivePlacement` during scaling events.
```

Author: Ack.

- Automated scaling within the JobSet controller (scaling should be triggered by external actors like HPA or Kueue).

## Proposal

### User Stories

#### Story 1
As a Kubeflow Training Operator v2 user, I want PyTorch elastic training that scales worker count during failures or when GPUs become available.

Currently JobSet rejects `parallelism`/`completions` changes needed by `elasticPolicy`. With this KEP:

```yaml
apiVersion: jobset.x-k8s.io/v1alpha2
kind: JobSet
metadata:
  name: pytorch-elastic
spec:
  replicatedJobs:
  - name: workers
    replicas: 1
    template:
      spec:
        completionMode: Indexed
        parallelism: 2      # Can scale: 2→8
        completions: 2
        template:
          spec:
            containers:
            - name: trainer
              image: pytorch/torchserve
              command: ["torchrun", "--nproc_per_node=1", "train.py"]
              resources:
                limits:
                  nvidia.com/gpu: 1
# kubectl patch jobset pytorch-elastic -p '{"spec":{"replicatedJobs":[{"replicas":1,"template":{"spec":{"parallelism":8,"completions":8}}}]}}'
```

#### Story 2

> As a Google Cloud TPU user, I want my training to continue when one TPU slice fails by scaling down replicas.

Author: My understanding is that while pure JAX historically didn't support dynamic scaling out of the box, modern orchestration layers (like Pathways) have introduced mechanisms to handle fault tolerance and dynamic resizing for JAX workloads when a TPU slice fails. That being said, I'll wait for @GiuseppeTT or @imreddy13 to confirm the current state of JAX elasticity.

TPU slices (16 nodes each) fail atomically. Currently entire JobSet fails. With elastic JobSets:

```yaml
apiVersion: jobset.x-k8s.io/v1alpha2
kind: JobSet
metadata:
  name: tpu-training
spec:
  replicatedJobs:
  - name: workers
    replicas: 4          # One Job per TPU slice
    template:
      spec:
        completionMode: Indexed
        parallelism: 16    # 16 TPU v4 cores per slice
        completions: 16
        template:
          spec:
            nodeSelector:
              cloud.google.com/gke-tpu-topology: 4x4
# TPU slice 2 fails → scale: replicas: 4→3 (controller deletes workers-3)
```

#### Story 3
As a cluster admin, I want JobSets to integrate with Kueue elastic quotas for dynamic GPU scaling.

Kueue cannot scale JobSets today. With elastic support, Kueue scales based on cluster utilization:

```
Cluster 100 GPUs @ 80% → Kueue scales: replicas 20→25
Cluster 100 GPUs @ 95% → Kueue scales: replicas 25→20
```

> ### Risks and Mitigations

Reviewer: Any risks with FailurePolicies or SuccessPolicies API?

> ### Risks and Mitigations

Author: Added some risks based on elastic Index Job risks

- **Resource Leakage:** Deleting a replica might leave orphaned pods if the child Job is not cleaned up properly.
  - *Mitigation:* The controller will rely on OwnerReferences and the standard Job deletion lifecycle to ensure all pods are terminated.
- **Index Fragmentation:** Scaling down and then up could lead to gaps in naming or indexing.
  - *Mitigation:* The controller will always scale down by removing the highest-indexed Jobs and scale up by appending new indices sequentially.

## Design Details

### API Details

The validation webhook for `jobset.x-k8s.io/v1alpha2` will be modified to allow updates to:

1. spec.replicatedJobs[i].replicas (Horizontal Job scaling)
2. spec.replicatedJobs[i].template.spec.parallelism (Internal Pod scaling)

> 3. spec.replicatedJobs[i].template.spec.completions (Internal Pod scaling)

Reviewer: You need to update goal + overview to say that we also allow to modify parallelism/completions.

	// Replicas is the number of desired Jobs for this ReplicatedJob.
	// This field is changed from immutable to mutable.
	Replicas int32 `json:"replicas"`

	// Template is now partially mutable.
  // Specifically, parallelism and completions can be updated
  // to support Elastic Indexed Jobs.
  Template batchv1.JobTemplateSpec `json:"template"`
}

```

### Implementation

The JobSet controller will implement a multi-path reconciliation logic to support both horizontal and internal elasticity. This ensures the JobSet state remains synchronized with both spec.replicas and the template.spec parameters.

#### Index Sorting:
The controller fetches all child Jobs owned by the JobSet and sorts them numerically by their index suffix (e.g., workers-0, workers-1).

#### Horizontal Scaling (Job-level):

Scale Up: If spec.replicas > actual Job count, the controller creates new Jobs for the missing indices sequentially (e.g., if scaling from 2 to 4, it creates indices 2 and 3).

Scale Down (LIFO): If actual Job count > spec.replicas, the controller deletes Jobs starting from the highest index downward. This preserves lower-indexed Jobs, which often host "leader" or "master" processes.

> #### Internal Scaling (Pod-level):

Reviewer: Shall we just name this:
- Horizontal Job Scaling (ReplicatedJob level)
- Horizontal Pod Scaling (Pod-Level)

Template Propagation: When the controller detects an update to spec.replicatedJobs[i].template.spec.parallelism or completions, it will iterate through all existing healthy child Jobs.

> Opportunistic Patching: The controller will issue a Patch request to each existing child Job to synchronize its spec.parallelism and spec.completions with the new values in the JobSet template.

Reviewer: Yeah, that makes sense, it looks like additional workers will wait for barrier to join the rendezvous: https://docs.pytorch.org/docs/stable/elastic/rendezvous.html
We should verify how it works in practice, if the JobSet controller fails during scale up, and only subset of Pods will join the training group.

Elastic Integration: By patching the child Jobs, the JobSet controller leverages the native Kubernetes Elastic Indexed Job functionality to scale the pod count within each job replica dynamically.

#### Failure Handling Priority

> To prevent "scaling into a fire," the controller will prioritize stabili  ty over elasticity:

Reviewer: ```suggestion
To prevent "scaling into a fire," the controller will prioritize stability over elasticity:
```

If any Job in the JobSet is in a Failed state, the controller will pause both horizontal and internal scaling operations until the FailurePolicy has reconciled the failed resource (either by restarting the set or hitting the backoffLimit).

#### Graceful Termination:
During any scale-down event (either removing a Job replica or reducing internal parallelism), the controller relies on the terminationGracePeriodSeconds of the Jobs and Pods. This allows applications like PyTorch Elastic to perform final checkpoints or clean shutdowns.

### Status Management

The JobSet.Status.ReplicatedJobStatus must reflect the scaling operation.

Ready/Active Counts: During a scale-down, the active count must be decremented as soon as the Job deletion is initiated.

> Conditions: A new condition Scaling may be added to the JobSet status to indicate a scale-up or scale-down is in progress.

Author: Thanks for confirming, @kannon92 

Does elastic Job rely on eventual consistency of the status fields instead?

If so, we can probably remove the condition for scaling and use the former instead.

  - *Mitigation:* The validating webhook will implement a strict block against late-stage mutations. If `jobset.status.conditions` contains a condition of type `Completed` or `Failed` with status `True`, any updates to `parallelism`, or `completions` will be rejected with an API validation error (e.g., `Forbidden: Cannot mutate a JobSet in a terminal state`).

## Design Details

### API Details

The validation webhook for `jobset.x-k8s.io/v1alpha2` will be modified to allow updates to:

1. spec.replicatedJobs[i].template.spec.parallelism (Horizontal Pod Scaling)
2. spec.replicatedJobs[i].template.spec.completions (Horizontal Pod Scaling)

```golang

type ReplicatedJob struct {
  // Name of the ReplicatedJob.
  Name string `json:"name"`

  // Replicas is the number of desired Jobs for this ReplicatedJob.
  // This field remains immutable in this iteration.
  Replicas int32 `json:"replicas"`

  // Template is now partially mutable.
  // Specifically, parallelism and completions can be updated
  // to support Elastic Indexed Jobs.
  Template batchv1.JobTemplateSpec `json:"template"`
}

```

### Implementation

> The JobSet controller will mplement reconciliation logic to support internal elasticity. This ensures the JobSet state remains synchronized with the template.spec parameters.

Reviewer: ```suggestion
The JobSet controller will implement reconciliation logic to support internal elasticity. This ensures the JobSet state remains synchronized with the template.spec parameters.
```

### Graduation Criteria

> To ensure stability for existing workloads, this feature will be introduced incrementally using a standard feature gate.

Author: Done

Furthermore, the number of worker pods within those individual Jobs may need to dynamically scale up or down based on cluster resource availability, requiring the underlying Job's `parallelism` and `completions` to be mutable as well.

### Goals

- Support scaling the number of replicas in a `ReplicatedJob` after the JobSet has been created.
- Leverage KEP-3715 (Elastic Indexed Job) to support internal scaling of child Jobs by allowing `parallelism` and `completions` to be mutated in tandem.
- Ensure the JobSet controller correctly manages the lifecycle (creation/deletion) of child Jobs based on the updated replica count.
- Maintain the stability of existing JobSet features like `DependsOn` and `ExclusivePlacement` during scaling events.

### Non-Goals

- Supporting mutability of the `JobTemplate` or container images mid-run.
- Automated scaling within the JobSet controller (scaling should be triggered by external actors like HPA or Kueue).

## Proposal

### User Stories

> #### Story 1

Reviewer: Ideally, we want to allow both for the long term. Replicas are useful for large-scale training jobs requiring over 700-1000 GPUs, since it will reduce reconciliation load on a single Job controller.
However, we didn't get a chance to integrate it with TrainJob yet: https://github.com/kubeflow/trainer/issues/2318 (cc @tenzen-y)

If we want, we can relax immutability only for `parallelism` and `completions` in the first iteration.
WDYT @aniket2405 @kannon92 @GiuseppeTT ?

> #### Story 1

Reviewer 2: That being said. I think we should add `replicas` mutability to a "Future work" section

        completions: 2
        template:
          spec:
            containers:
            - name: trainer
              image: pytorch/torchserve
              command: ["torchrun", "--nnodes=2:8", "--nproc_per_node=1", "train.py"]
              resources:
                limits:
                  nvidia.com/gpu: 1
# kubectl patch jobset pytorch-elastic -p '{"spec":{"replicatedJobs":[{"replicas":1,"template":{"spec":{"parallelism":8,"completions":8}}}]}}'
```

#### Story 2
As a Google Cloud TPU user, I want my training to continue when one TPU slice fails by scaling down replicas.

TPU slices (16 nodes each) fail atomically. Currently entire JobSet fails. With elastic JobSets:

```yaml
apiVersion: jobset.x-k8s.io/v1alpha2
kind: JobSet
metadata:
  name: tpu-training
spec:
  replicatedJobs:
  - name: workers
    replicas: 4          # One Job per TPU slice
    template:
      spec:
        completionMode: Indexed
        parallelism: 16    # 16 TPU v4 cores per slice
        completions: 16
        template:
          spec:
            nodeSelector:
              cloud.google.com/gke-tpu-topology: 4x4
# TPU slice 2 fails → scale: replicas: 4→3 (controller deletes workers-3)
```

> #### Story 3

Reviewer 2: My understanding of this use case is that Kueue needs to be able to react to changes to `jobSet.spec.replicatedJobs[].replicas`. But is there a use case to make `jobSet.spec.replicatedJobs[].replicas` mutable? Perhaps from PyTorch elastic training + Kubeflow trainer?

  - *Mitigation:* The JobSet controller will not attempt to reset or mask the underlying `status.Failed` count of child Jobs; users must configure their `FailurePolicy` with the understanding that scale-down events do not forgive previous pod failures.
- **Race Conditions on Job Completion:** Changing completion semantics dynamically could conflict with the Job controller declaring a child Job finished.
  - *Mitigation:* This is handled upstream by the native Job controller, which uses `Update` when setting the finished status (triggering object modified errors on concurrent spec changes). Additionally, `spec.completions` is strictly immutable once a Job reaches a terminal condition, preventing the JobSet controller from improperly modifying finished Jobs.

#### JobSet Policy & State Management
- **Scale-down Deletion Triggering FailurePolicy:** When horizontal scaling reduces `replicas`, the JobSet controller must delete the highest-indexed child Jobs. The reconciliation loop might misinterpret this deletion as an unexpected Job failure.

>   - *Mitigation:* The controller logic will explicitly differentiate between "Jobs deleted due to a scale-down event" and "Jobs that failed." Controlled deletions triggered by a `spec.replicas` update will bypass `FailurePolicy` evaluation.

Reviewer 2: I like that this is said explicitly in the KEP so please keep it, but the implementation might not need to worry about this.

tl;dr: deleting healthy Jobs typically do not fail them. Currently, the JobSet controller only restarts when a Job fails, if it is "healthy deleted", the JobSet controller just recreates it.

Author: Acknowledged. I made this clearer in the mitigation.

  - *Mitigation:* The validating webhook will implement a strict block against late-stage mutations. If `jobset.status.conditions` contains a condition of type `Completed` or `Failed` with status `True`, any updates to `replicas`, `parallelism`, or `completions` will be rejected with an API validation error (e.g., `Forbidden: Cannot mutate a JobSet in a terminal state`).

## Design Details

### API Details

The validation webhook for `jobset.x-k8s.io/v1alpha2` will be modified to allow updates to:

1. spec.replicatedJobs[i].replicas (Horizontal Job scaling)
2. spec.replicatedJobs[i].template.spec.parallelism (Horizontal Pod Scaling)
3. spec.replicatedJobs[i].template.spec.completions (Horizontal Pod Scaling)

```golang

type ReplicatedJob struct {
	// Name of the ReplicatedJob.
	Name string `json:"name"`

	// Replicas is the number of desired Jobs for this ReplicatedJob.
	// This field is changed from immutable to mutable.
	Replicas int32 `json:"replicas"`

	// Template is now partially mutable.
  // Specifically, parallelism and completions can be updated
  // to support Elastic Indexed Jobs.
  Template batchv1.JobTemplateSpec `json:"template"`
}

```

> ### Implementation

Reviewer: Makes sense, we can have a feature flag for some time.

> ### Implementation

Reviewer 3: @GiuseppeTT 
Are you thinking of a feature gate  to iterate on alpha and add phases?

Or to control the behavior for users who don't want this behavior?

I'm fine with a feature gate but just want to clarify your ask.

#### Graceful Termination:
During any scale-down event (either removing a Job replica or reducing internal parallelism), the controller relies on the terminationGracePeriodSeconds of the Jobs and Pods. This allows applications like PyTorch Elastic to perform final checkpoints or clean shutdowns.

### Status Management

The JobSet.Status.ReplicatedJobStatus must reflect the scaling operation.

Ready/Active Counts: During a scale-down, the active count must be decremented as soon as the Job deletion is initiated.

Conditions: A new condition Scaling may be added to the JobSet status to indicate a scale-up or scale-down is in progress.

### Defaulting/Validation

- **Mutable Fields:** `replicas`, `template.spec.parallelism`, and `template.spec.completions`.
- **Minimum Value Constraints:** All three mutable scaling fields (`replicas`, `parallelism`, and `completions`) must be strictly `>= 1`. Setting `parallelism` to `0` natively suspends a Job, which conflicts with JobSet's explicit top-level `suspend` field. Elasticity cannot be used as a suspension mechanism.

> - **Elastic Indexed Job Constraints:** Horizontal pod-level scaling inherits the strict validation rules of upstream Elastic Indexed Jobs:

Reviewer 2: Keep in mind that JobSet currently doesn't fully validate the Job template. So it is currently possible to create a JobSet with an invalid JobTemplate and get it stuck because it can't create Jobs.

The implementation should validate the changes manually.

- **Topology & Placement:**
  - Test scaling in a JobSet that uses `ExclusivePlacement` to ensure pods are still placed correctly on new nodes.
- **Policy & State Interactions (New):**
  - Verify that scaling down a `ReplicatedJob` (deleting a child Job) does *not* trigger the JobSet's `FailurePolicy`.
  - Verify that if a child Job fails, the `FailurePolicy` evaluates and executes (e.g., restarting the JobSet) and correctly aborts or defers any pending scaling operations.

### Implementation History
Mar 21, 2024: Issue #463 opened as an RFC.

Jul 30, 2024: Discussions regarding framework requirements for PyTorch v2.

Feb 04, 2026: KEP formally proposed for v1alpha2.

### Drawbacks
Application Awareness: This feature is only useful if the application running inside the JobSet can handle its membership changing dynamically.

Controller Overhead: Frequent scaling updates could increase the load on the JobSet controller and the Kubernetes API server.

### Alternatives
HorizontalPodAutoscaler (HPA) on Jobs: HPA can scale the parallelism of a standard Job, but it cannot manage the multi-job coordination that JobSet provides (e.g., DNS, multi-interface networking).

> Manual Re-creation: Deleting the JobSet and creating a new one with the desired size. This causes a full restart of all workers, which is inefficient for large-scale training.

Author: I've added a 'Future Work' section to capture this.

Controller Overhead: Frequent scaling updates could increase the load on the JobSet controller and the Kubernetes API server.

## Alternatives
HorizontalPodAutoscaler (HPA) on Jobs: HPA can scale the parallelism of a standard Job, but it cannot manage the multi-job coordination that JobSet provides (e.g., DNS, multi-interface networking).

Manual Re-creation: Deleting the JobSet and creating a new one with the desired size. This causes a full restart of all workers, which is inefficient for large-scale training.

## Future Work

While this initial implementation focuses strictly on Pod-level scaling (`parallelism` and `completions`), the ultimate goal is to support full elasticity, including Job-level scaling. The following items capture the design ideas, user stories, and identified risks for these future iterations.

### 1. Horizontal Job Scaling (`replicas` mutability)
Future iterations will introduce mutability to `jobSet.spec.replicatedJobs[].replicas`. This will allow the JobSet controller to natively handle scaling down (e.g., in response to a hardware failure) and scaling up (when new nodes become available) without requiring a full teardown.

> #### Motivating User Story: TPU Slice Failures

Author: Done
